pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -229,14 +292,45 @@ func (r *remoteReadQueryRequest) WithQuery(_ string) (MetricsQueryRequest, error
 	panic("not implemented")
 }
 
-func (r *remoteReadQueryRequest) WithHeaders([]*PrometheusHeader) MetricsQueryRequest {
+func (r *remoteReadQueryRequest) WithHeaders(_ []*PrometheusHeader) MetricsQueryRequest {
 	panic("not implemented")
 }
 
-func (r *remoteReadQueryRequest) WithStartEnd(_ int64, _ int64) MetricsQueryRequest {
-	panic("not implemented")
+// WithStartEnd clones the current remoteReadQueryRequest with a new start and end timestamp.
+func (r *remoteReadQueryRequest) WithStartEnd(start int64, end int64) MetricsQueryRequest {
+	// TODO handle error (we need to return it from this function, but the changes required in other places are very annoying)
+	clonedQuery, _ := cloneRemoteReadQuery(r.query)
+	clonedQuery.StartTimestampMs = start
+	clonedQuery.EndTimestampMs = end
+
+	if clonedQuery.Hints != nil && clonedQuery.Hints.StartMs != 0 {
+		clonedQuery.Hints.StartMs = start
+	}
+	if clonedQuery.Hints != nil && clonedQuery.Hints.EndMs != 0 {
+		clonedQuery.Hints.EndMs = end
+	}
+
+	clonedReq := *r
+	clonedReq.query = clonedQuery
+	return &clonedReq
```

Comment:
this feels off, couldn't we just update the r.query and convert that into the result with `remoteReadToMetricsQueryRequest`

Reply:
I see your comment and I also don't like much this function, but the `WithXXX()` functions are expected to return a deep copy without manipulating any of the original request data (that's by contract of these functions). If we do manipulate `r.query` then we're going to break the contract.

E.g. we take a similar approach in `PrometheusRangeQueryRequest` too, but there we don't have a complex nested struct like `prompb.Query`:
https://github.com/grafana/mimir/blob/5571e24d8da9ec875a5834fc6ce57829a608f2f4/pkg/frontend/querymiddleware/model_extra.go#L158-L168

Reply:
Discussed offline. I've changed the code to use `remoteReadToMetricsQueryRequest()`.